components/net/indexeddb/engines/sqlite.rs

```diff
@@ -0,0 +1,513 @@
+
+    pub fn encode(key: &IndexedDBKeyType) -> Vec<u8> {
+        bincode::serialize(&key).unwrap()
+    }
+
+    pub fn decode(s: &[u8]) -> Option<IndexedDBKeyType> {
+        bincode::deserialize(s).ok()
+    }
+}
+
+// These pragmas need to be set once
+const DB_INIT_PRAGMAS: [&str; 2] = ["PRAGMA journal_mode = WAL;", "PRAGMA encoding = 'UTF-16';"];
+
+// These pragmas need to be run once a connection.
+const DB_PRAGMAS: [&str; 4] = [
+    "PRAGMA synchronous = NORMAL;",
+    "PRAGMA journal_size_limit = 67108864 -- 64 megabytes;",
+    "PRAGMA mmap_size = 67108864 -- 64 megabytes;",
+    "PRAGMA cache_size = 2000;",
+];
+
+fn range_to_query(range: IndexedDBKeyRange) -> Condition {
+    // Special case for optimization
+    if let Some(singleton) = range.as_singleton() {
+        let encoded = serialize::encode(singleton);
+        return object_data_model::Column::Data.eq(encoded).into_condition();
+    }
+    let mut parts = vec![];
+    if let Some(upper) = range.upper.as_ref() {
+        let upper_bytes = serialize::encode(upper);
+        let query = if range.upper_open {
+            object_data_model::Column::Data.lt(upper_bytes)
+        } else {
+            object_data_model::Column::Data.lte(upper_bytes)
+        };
+        parts.push(query);
+    }
+    if let Some(lower) = range.lower.as_ref() {
+        let lower_bytes = serialize::encode(lower);
+        let query = if range.upper_open {
+            object_data_model::Column::Data.gt(lower_bytes)
+        } else {
+            object_data_model::Column::Data.gte(lower_bytes)
+        };
+        parts.push(query);
+    }
+    let mut condition = Condition::all();
+    for part in parts {
+        condition = condition.add(part);
+    }
+    condition
+}
+
+macro_rules! err {
+    ($e:expr) => {
+        Err(format!("{:?}", $e))
+    };
+}
+
+pub struct SqliteEngine {
+    db_path: PathBuf,
+    connection: DatabaseConnection,
+}
+
+impl SqliteEngine {
+    pub fn new(base_dir: &Path, db_info: &IndexedDBDescription, version: u64) -> Self {
+        let mut db_path = PathBuf::new();
+        db_path.push(base_dir);
+        db_path.push(db_info.as_path());
+        db_path.push("db.sqlite");
+
+        let connection = if db_path.exists() {
+            HANDLE.block_on(async { Self::get_connection(&db_path).await })
+        } else {
+            std::fs::create_dir_all(db_path.parent().unwrap()).unwrap();
+            std::fs::File::create(&db_path).unwrap();
+            HANDLE.block_on(async { Self::init_db(&db_path, db_info, version).await })
+        }
+        .unwrap();
+
+        for stmt in DB_PRAGMAS {
+            HANDLE
+                .block_on(connection.execute_unprepared(stmt))
+                .unwrap();
+        }
+
+        Self {
+            connection,
+            db_path,
+        }
+    }
+
+    async fn get_connection(path: &Path) -> Result<DatabaseConnection, DbErr> {
+        Database::connect(format!("sqlite://{}", path.display())).await
```

Comment:
It looks odd to await here, since we immediately return and the caller calls await. Can we remove this await?